packages/wallet/src/services/SmartTxSubmitProvider.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  Cardano,
+  HealthCheckResponse,
+  ProviderError,
+  ProviderFailure,
+  SubmitTxArgs,
+  TxSubmitProvider,
+  cslUtil
+} from '@cardano-sdk/core';
+import { ConnectionStatus, ConnectionStatusTracker } from './util';
+import { Observable, combineLatest, filter, firstValueFrom, from, mergeMap, take, tap } from 'rxjs';
+import { RetryBackoffConfig, retryBackoff } from 'backoff-rxjs';
+
+export interface RetryingTxSubmitProviderProps {
+  retryBackoffConfig: RetryBackoffConfig;
+}
+
+export type TipSlot = Pick<Cardano.Tip, 'slot'>;
+
+export interface RetryingTxSubmitProviderDependencies {
+  txSubmitProvider: TxSubmitProvider;
+  tip$: Observable<TipSlot>;
+  connectionStatus$: ConnectionStatusTracker;
+}
+
+/**
+ * Wraps a `TxSubmitProvider` to enchance it's `submitTx` with the following functionality:
+ * - Immediately rejects if network tip is already >= `ValidityInterval.invalidHereafter`
+ * - Awaits for the following conditions before submitting:
+ *   - Network tip is ahead of tx body `ValidityInterval.invalidBefore`.
+ *   - Connection status is 'up'.
+ * - Re-submits transactions that failed to submit due to connection or recoverable provider issue.
+ */
+// REVIEW: an alternative name is ManagedTxSubmitProvider.
```

Comment:
I'm happy with `SmartTxSubmitProvider` now I read the code :-)

Reply:
3af7e3498ba6f0a20cc1fa52c410a4a7dec9ea0a